File: C++ files/HashTable - C++/hashtable_separate_chaining.cpp
/*
Name: Minh Q. Vu

This python file is a part of my effort in getting myself refreshed with Data Structure and Algorithms concepts so
I can later tackle Leetcode challenges with more confidence. 

========================================================= Hash Table with Seperate Chaining =========================================================



*/ 
    
#include "hashtable_separate_chaining.h" 

// Constructor for a new Hash Table with seperate chaining
HashTable_SeparateChaining::HashTable_SeparateChaining() 
{
    this->t = new HashTableEntry_SeparateChaining * [table_size_sc]; // an array of pointers 
    for (int i = 0; i < table_size_sc; i++) {
        t[i] = NULL; 
    }
};


// Destructor for a new Hash Table
HashTable_SeparateChaining::~HashTable_SeparateChaining()
{
    for (int i = 0; i < table_size_sc; i++){
        delete t[i]; 
    }
    delete[] t; 
}; 


// Hash Function
int HashTable_SeparateChaining::hashFunc(int key)
{
    return key % table_size_sc; 
};


/** Insert new pair of key-value 
Move to the bucket corresponds to the above calculated hash index and sequentially search for a possible duplicate key.  
If there is none then insert the new node at the end of the list.
*/
void HashTable_SeparateChaining::insert(int key, int val)
{
    int h = hashFunc(key); 

    if (t[h] == NULL) {
        t[h] = new HashTableEntry_SeparateChaining(key, val); 
    } else {
        HashTableEntry_SeparateChaining* temp = t[h]; 
        HashTableEntry_SeparateChaining* prev = nullptr; 
        //sequentially search for possible duplicate key
        while (temp != NULL){
            if (temp->k == key){
                temp->v = val; 
                return; 
            }
            prev = temp; 
            temp = temp->next; 
        }
        // we reach the end of the linked list
        prev->next = new HashTableEntry_SeparateChaining(key, val); 
    }
}


// Search a pair
int HashTable_SeparateChaining::searchKey(int key)
{
    int h = hashFunc(key);

    HashTableEntry_SeparateChaining* temp = t[h]; 
    while (temp && temp->k != key) {
        temp = temp->next; 
    }

    if (!temp) return -1;
    else return temp->v; 
}


// Remove a pair
void HashTable_SeparateChaining::remove(int key)
{
    int h = hashFunc(key); 
    
    HashTableEntry_SeparateChaining* prev = nullptr; 
    HashTableEntry_SeparateChaining* temp = t[h]; 

    // special case: t[h] is NULL
    if (!temp) {
        return;
    } 

    // search for desired key
    while (temp && temp->k != key) {
        prev = temp;
        temp = temp->next; 
    }

    if (!temp) {                        // cannot find the desired key
        return; 
    } else {
        if (!temp->next && !prev) {     // special case 1: when t[h] only has 1 node and it has the desired key
            t[h] = NULL; 
        } else {    
            if (!prev) {                // special case 2: when remove the head of a long list
                t[h] = temp->next; 
            } else {                    // normal case: remove a node in the middle of the list
                prev->next = temp->next; 
            }
        }

        delete temp; 
    }
};



// Print all pair of key-value
void HashTable_SeparateChaining::printTable()
{
    for (int i = 0; i < table_size_sc; i++){
        std::cout << "Slot index: " << i << "       ";
        if (t[i] == NULL) {
            std::cout << "NULL" << std::endl; 
        } else {
            HashTableEntry_SeparateChaining* temp = t[i]; 
            while (temp != NULL){    
                std::cout << "Key: " << temp->k << " -- ";
                std::cout << "Value: " << temp->v << " ---> ";
                temp = temp->next; 
            }
            std::cout << "NULL\n"; 
        }
    }
}; 

File: Python files/HashTable - Python/hash_table_separate_chaining.py
"""
Name: Minh Q. Vu

This python file is a part of my effort in getting myself refreshed with Data Structure and Algorithms concepts so
I can later tackle Leetcode challenges with more confidence. 

========================================================= Hash Table =========================================================
Hash tables are a type of data structure in which the address or the index value of the data element is generated 
from a hash function. That makes accessing the data faster as the index value behaves as a key for the data value. 
In other words Hash table stores key-value pairs but the key is generated through a hashing function.

So the search and insertion function of a data element becomes much faster as the key values themselves become the
index of the array which stores the data.

🧠 Key Concepts Preserved
- Each bucket is a linked list of key-value pairs.
- Handles collisions via separate chaining.
- Supports insert, search, remove, and print operations.

"""

TABLE_SIZE_SC = 10

# Same as a LinkedList node
class HashTableEntry:
    def __init__(self, key: int, value: int):
        self.k = key
        self.v = value
        self.next = None


class HashTableSeparateChaining:
    def __init__(self):
        # Initialize table with empty slots (None)
        self.t = [None for _ in range(TABLE_SIZE_SC)]

    def hash_func(self, key: int) -> int:
        return key % TABLE_SIZE_SC

    def insert(self, key: int, value: int) -> None:
        h = self.hash_func(key)

        if self.t[h] is None:
            self.t[h] = HashTableEntry(key, value)
        else:
            temp = self.t[h]
            prev = None

            # Search for duplicate key
            while temp:
                if temp.k == key:
                    temp.v = value  # Update existing key
                    return
                prev = temp
                temp = temp.next

            # Append new node at the end
            prev.next = HashTableEntry(key, value)

    def search_key(self, key: int) -> int:
        h = self.hash_func(key)
        temp = self.t[h]

        while temp and temp.k != key:
            temp = temp.next

        return temp.v if temp else -1

    def remove(self, key: int) -> None:
        h = self.hash_func(key)
        temp = self.t[h]
        prev = None

        if not temp:
            return  # Nothing to remove

        while temp and temp.k != key:
            prev = temp
            temp = temp.next

        if not temp:
            return  # Key not found

        if not prev:
            # Remove head node
            self.t[h] = temp.next
        else:
            # Remove middle or tail node
            prev.next = temp.next

    def print_table(self) -> None:
        for i in range(TABLE_SIZE_SC):
            print(f"Slot index: {i}       ", end="")
            if self.t[i] is None:
                print("NULL")
            else:
                temp = self.t[i]
                while temp:
                    print(f"Key: {temp.k} -- Value: {temp.v} ---> ", end="")
                    temp = temp.next
                print("NULL")







